posts/coding-intentionally-in-bash-grains.md

```diff
@@ -0,0 +1,185 @@
+
+It is a clever solution, but what is the code telling us?  That hexadecimal is important here?  That the problem fundamentally revolves around 16?  After re-reading the problem statement, it's pretty clear that neither of those are the case.  The student and I brainstormed some ideas for how to more clearly communicate intent.  Here are a few things we came up with:
+
+### First option: binary
+
+Because we've got a bunch of things doubling (and thus, a bunch of powers of 2), let's look at what's happening in binary to see if that will help us.
+
+---
+
+The first square has 1 grain on it.  In binary, this would also be `0b1` (the `0b` just meaining "this is a binary number", with the actual number being `1`).
+
+The second square has 2 grains on it.  In binary, `0b10`.  The total so far is 3 (or `0b11`).
+
+The third square has 4 (`0b100`) grains on it.  Total so far: 7 (`0b111`).
+
+The fourth square has 8 grains (`0b1000`).  Total so far: 15 (`0b1111`).
+
+---
+
+Can you see the pattern?
+
+Each square represents another binary digit, and adding all of them together just makes a bunch of 1's.
+
+In the student's solution, we could replace the F's with 64 1's (one for each square)!
+
+```bash
+bc <<< "ibase=2;1111111111111111111111111111111111111111111111111111111111111111"
+```
+
+More intentional, because it more closely matches what the problem gives us.  But, with it being so long and with us not speaking robot, a long, essentially uncountable string of 1's is maybe not an improvement.
+
+### Second option: the brute-force calculation
+
+OK, so maybe we abandon the non-decimal numbering systems altogether.  Why don't we make the code match how we would total up the number of grains on a chessboard by hand, by counting the grains on each square?
+
+```bash
+total=0
+current_grains=1
+
+for square in {1..64}; do
+  total=$( bc <<< "$total + $current_grains" )
+  current_grains=$( bc <<< "$current_grains * 2" )
+done
+
+echo "$total"
+```
+
+That's a lot more readable and understandable.  The code clearly shows that the number of squares on the chess board is a driving factor, as well as the doubling effect for each square.  I think this is better than the initial solution.
+
+However.
+
+It's slow.  Looping, adding, and repeatedly calling out to an external command?  They all stack up to a slow-ish run-time.  Now, is that a huge deal?  No.  If you're scripting this in Bash, you've probably already decided you don't have speed constraints.  But could it be better?  Yes.
+
+### Third option: direct calculation
+
+In the math biz (that's what cool math folks call it), we've got a name for the process "take something and then multiply it by the same number a bunch of times": exponents.
+
+In fact, the number of grains on each square can be expressed as:
+
+```bash
+bc <<< "2^($square - 1)"
+# We subtract one from $square because square 1 has 1 (2^0) grains on it.
```

Comment:
I happen to know why it works, but I know that when I read something like this in a potentially scary domain and someone says "you can do X" I kind of feel dumb when I don't know why.

Since you're explaining below, I think we could make this more impactful (and less intimidating) by having the explanation come first, and then the formula as the conclusion.